### ISSUE TITLE
Frontend performance

### ISSUE THREAD

**Issue by mikusp** [AccelerateHS/accelerate#278]
We are trying to improve performance of our application that uses Accelerate on CUDA backend. Recently we came to the realization, that frontend phase of Accelerate is slowing us down. That's because of our use case - under the scenes, user builds a computation, that has to be compiled, computed, and the result viewed on the screen. In other words, latency (time between requesting the result to receiving it) is more important than throughput (performance of the generated kernels). Using run1 family of functions is not feasible, as it would cause a major redesign of our app, and most of the computations are only run once.

I prepared a simple test program that uses run exactly once and generates three CUDA kernels. I cannot share the source code nor the binary with you.
When I run the program with +RTS -s -p options, it completes in about 2.5s.
Profiling reveals, that convertAccWith function is responsible for 94% of inherited time. Whole .prof file is available here: https://gist.github.com/mikusp/292f47c39847ffb98d79

I thought that disabling optimizations in accelerate by passing flags -fno-simplify, -fno-fusion, etc. will speed up things, but I didn't see a measurable difference.

Finally, I tried to optimize the biggest individual offender, that is idxToInt. Patch to accelerate is here: https://gist.github.com/mikusp/46913963c813175cfb3f accelerate-cuda requires two trivial changes to compile after applying this patch. Result is pretty satisfying, because now the test program runs in ~1.8s.

Please let us know if you have any thoughts about frontend performance. Maybe there are some ways to sacrifice kernel performance for compilation speed or other low hanging fruits that would be easy to optimize. I can provide you with more logs/prof files/examples if it'd be helpful.

**Comment by tmcdonell** [comment]
Is the idea with the patch to keep a real `Int` as part of the `SuccIdx` constructor, that would have been the result of running `idxToInt`? To be honest, that seems like a bad idea --- it too easily gets out of sync, and now you'll have silent errors in the generated code that could be very difficult to track down.

I have a different program that takes a lot of time in the front end, but that one doesn't flag `idxToInt` as being particular problematic, so they may be different issues. Anyway, I am aware that the front end can take a long time.

If you can run your program with `-ddump-simpl-stats` and let me know how many kernels get generated, that might be helpful.

**Comment by mikusp** [comment]
Yes, that was the idea. I agree that it can get out of sync. However, the only functions that touch this Int are smart constructor succIdx and idxToInt, and only succIdx modifies it. It's impossible to accidentally mix SuccIdx and succIdx in code, as they have different number of arguments. To somehow minimise the risk, SuccIdx could be hidden from other modules and a pattern synonym that ignores the Int parameter exported, for pattern matching on Idx.

```
$ ./test1 +ACC -ddump-simpl-stats -ACC
2.544:Total ticks: 73968

143 Inline
  64 prj/Tuple
  48 prj/Var
  24 prj/Let
  7 Var
379 RuleFired
  227 CSE
  32 commutes *
  18 -y+x
  16 x*0
  16 x*1
  12 commutes /=*
  11 generateD
  10 mapD
  8 aletD/eliminate
  8 commutes .&.
  6 aletD/float
  4 commutes +
  4 x+0
  4 x-x
  2 aletD/bind
  1 unzipD
68 BetaReduce
  55 inline exp
  13 dead exp
73281 Substitution
  72999 weakenE
  97 shrink exp
  89 inline
  34 weaken
  26 compose
  14 sink1
  13 replaceE/!
  8 replaceE/shape
  1 substitute
97 SimplifierDone 97
```

Two kernels are being generated (not three, as I wrote before).

**Comment by tmcdonell** [comment]
#272
